modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js

```diff
@@ -179,6 +179,14 @@ const CreationMenu = ({
 	);
 };
 
+const Item = ({children, href, onClick}) => {
+	return (
+		<ClayDropDown.Item href={href} onClick={onClick}>
+			{children}
+		</ClayDropDown.Item>
+	);
+};
+
```

Comment:
@julien We should be able to reduce duplication even further, if we pass `item` prop. If we have

```js
const Item = ({item, key}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                    key={key}
                    onClick={(event) => {
                        onCreationMenuItemClick(event, {item});
                    }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```
we can invoke it from two places with something like
```js
<Item item={item} key={index} />
```

I might be missing something that does not let us do this :thinking:

Reply:
You're right that does look even better. 

I see that you're suggesting passing a `key` prop, and in my understanding (which might be wrong) `key` is a special prop (like `ref`) so based on the code from above, what do you you think about a slight change?  

```js
const Item = ({item}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                    onClick={(event) => {
                        onCreationMenuItemClick(event, {item});
                    }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And then in those two places where it's used do

```js
<Item item={item} key={index} />
```

I know it seems strange to pass a key to the `Item` component (without it being used on declared inside that component) but React uses this internally. [This article](https://reactjs.org/warnings/special-props.html) or rather paragraph touches a bit on the topic, but I'd need to read more to actually know all the details.

To go a bit further, I also think that we have to "pass down" the `onCreationItemMenuClick` handler (but correct me if I'm wrong), because it's passed from `ManagementToolbar` > `CreationMenu` > `ItemList` > `Item` , which is going to make the code look like this (at least in my head) 

```javascript
const Item = ({item, onCreationMenuItemClick}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                    onClick={(event) => {
                        onCreationMenuItemClick(event, {item});
                    }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

```js
<Item item={item} key={index} onCreationMenuItemClick={onCreationMenuItemClick} />
```

Looks a bit ugly right?

So we could then say: let's pass this as `onClick` and use `...otherProps` in our Item component. 

```js
const Item = ({item, ...otherProps}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   {...otherProps}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And ... 

```js
<Item
     item={item} 
     key={index} 
     onClick={(event) => {
         onCreationMenuItemClick(event, {item});
    }} 
/>
```

It might not be the right time to do yet something else that would avoid passing down these props which is using `useContext` ... but I think we could consider it for the near future, what do you think?

Reply:
@julien 

You are very much right on `key`, nice catch :+1: 

`onCreationMenuItemClick` will not be available if we put it in file root. But, we can add it as an internal component, like we are doing with [ItemWithLabel here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/js/components/select-objects/SelectObjects.es.js#L111-L133). In fact, we can also do this with `ItemList` and remove its props.

Placing components in root implies they are separate components, only wired through props. This would makes sense if they are imported and used directly, like `ClayDropDown.ItemList` or `ClayDropDown.Item`. But, in our case, `ItemList` and `Item` are just internal helper components, used to squash duplication.

In the same way, `...otherProps` would make sense if `<Item>` was a generic component and we want flexible API. We have it in Clay in most or all components, and it's a good pattern. But, if `<Item>` is an internal component, we control 100% of use cases and there is no need for it.

You mentioned in Slack a different usage case where we don't support item icons and we should. This is an opportunity to also add this:
```js
<ClayDropDown.Item 
     symbolLeft={item.icon}
     ...
>
```

Reply:
> onCreationMenuItemClick will not be available if we put <Item> in file root. But, we can add <Item> as an internal component, like we are doing with ItemWithLabel here. In fact, we can also do this with ItemList and remove its props.

Hey @markocikos , sorry for not being clear enough.

When you say **in file root**, do you mean putting `Item` in it's own file?

I'm mostly asking because my point was that adding this:

```js
const Item = ({item}) => {
	return (
		<ClayDropDown.Item 
			href={item.href}
			onClick={(event) => {
				onCreationMenuItemClick(event, {item});
			}}
			>
			{item.label}
		</ClayDropDown.Item>
	);
};
```

and then adding it `ItemList`: 

```js
const ItemList = ({
	onCreationMenuItemClick,
	primaryItems,
	secondaryItems,
	visibleItemsCount,
}) => {
	let currentItemCount = 0;

	return (
		<ClayDropDown.ItemList>
			{primaryItems?.map((item, index) => {
				currentItemCount++;

				if (currentItemCount > visibleItemsCount) {
					return false;
				}

				return (
					<Item item={item} key={index} />
				);
			})}

			{secondaryItems?.map((secondaryItemsGroup, index) => (
				<ClayDropDown.Group
					header={secondaryItemsGroup.label}
					key={index}
				>
					{secondaryItemsGroup.items.map((item, index) => {
						currentItemCount++;

						if (currentItemCount > visibleItemsCount) {
							return false;
						}

						return (
							<Item item={item} key={index} />
						);
					})}

					{secondaryItemsGroup.separator && (
						<ClayDropDown.Item className="dropdown-divider" />
					)}
				</ClayDropDown.Group>
			))}
		</ClayDropDown.ItemList>
	);
};
```

In [CreationMenu,js](https://github.com/liferay/liferay-portal/blob/9272e31bc0297f9ad114de2c3a67f68aec8ea3cd/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/CreationMenu.js) will not "work". It might compile correctly, but at runtime, clicking on the `Item` will give you a runtime error like this

![](https://user-images.githubusercontent.com/5572/107641111-7c74cb80-6c73-11eb-8f81-af270dd9c30c.png)

Concerning using other `...otherProps`, I know that this is component we control at 100%, but I think using React's common patterns still makes sense, i.e. Since we have to pass the `onCreationMenuItemClick` callback anyway, I'd rather use this 

```js
const Item = ({item, ...otherProps}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   {...otherProps}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And  

```js
<Item
     item={item} 
     key={index} 
     onClick={(event) => {
         onCreationMenuItemClick(event, {item});
    }} 
/>
```

Rather than 

```js
const Item = ({item, onCreationMenuItemClick}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   onClick={(event) => {
                       onCreationMenuItemClick(event, {item});
                   }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

And 

```js
<Item item={item}  key={index} onCreationMenuItemClick={onCreationMenuItemClick} />
```

But my at the same time I want **everybody** (on the team not on the planet :joy:) to be OK with that, because I think it's important.

Concerning "icon support": good call! I think it's pretty trivial so I can add that too and we'll kill two birds with one stone.

Reply:
You can view this [here](https://github.com/liferay-frontend/liferay-portal/pull/774/commits/f0a2180df46063aad30f5fca0447b26f566eb812) or if we prefer 

```js
const Item = ({item, onCreationMenuItemClick}) => {
	return (
		<ClayDropDown.Item 
                    href={item.href}
                   symbolLeft={item.icon}
                   onClick={(event) => {
                       onCreationMenuItemClick(event, {item});
                   }}
                >
                        {item.label}
		</ClayDropDown.Item>
	);
};
```

I can always push another commit on top, and squash before forwarding

Reply:
@julien Everything you are suggesting makes sense if we are placing the `Info` component in file root. You are doing this in this PR:
```js
const CreationMenu = ({...}) => {
    return (
        <Item ... /> // we need to add `onCreationMenuItemClick` etc.
    )
}

const Item = ({...}) => {
    return (...)
}
```
Both are in the root of the `CreationMenu.js` file.

What I am suggesting is to make `Item` an internal private component:
```js
const CreationMenu = ({...}) => {
    const Item = ({...}) => {
        return (...)
    }

    return (
        <Item ... /> // we only need `item` prop
    )
}
```
[ItemWithLabel](https://github.com/liferay/liferay-portal/blob/master/modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/js/components/select-objects/SelectObjects.es.js#L111-L133) is an example where we are doing this.

The placement in root vs inside has many implications, as I mentioned in https://github.com/liferay-frontend/liferay-portal/pull/774#discussion_r574402978. `otherProps` makes sense to me in former, but not the latter. I think that, for our purpose, it makes more sense to place both `Item` and `ItemList` inside.

Reply:
@markocikos OK got it, now it all makes more sense for me too :+1: 

In this case and as you know I don't like the fact that the `ItemList` and `Item` function are going to be re-declared each time `CreationMenu` is going to be invoked, but OK, I'll do that.

Reply:
@markocikos see [
fd50016](https://github.com/liferay-frontend/liferay-portal/pull/774/commits/fd5001602931bab69656dab310a6fc0d8809faa3) and tell me if that's OK when you can - thanks. I'll squash before forwarding.

Reply:
I'm not sure, maybe there is an issue here :thinking: 

Summoning @wincent ! Is it safe to use React components within components? It makes sense from separation of concerns perspective, but is there a performance or a React quirk issue here?

Reply:
@markocikos just so you know, @wincent will be on PTO tomorrow, so if this is not a big risk (and I think it isn't) we could probably pass this as is (if you agree) and discuss this later on ... if it needs to be changed then we can do it in another PR - What are your thoughts on that? Thanks.

Reply:
@julien Yeah I agree. A performance issue is not urgent, if there is an issue with it we can iterate later on. And if there is a quirk with Reach that prevents this from working, we would see it already.

Let's push this! :rocket:

Reply:
> Summoning @wincent ! Is it safe to use React components within components? It makes sense from separation of concerns perspective, but is there a performance or a React quirk issue here?

tl;dr: Rule of thumb — prefer to hoist components up to the top-level when you can (for performance, to reduce GC pressure, to make clearer what the responsibility of each component is) _but_ as soon as there is a reason to inline the component (eg. because it'll be way more convenient to access state via a closure instead of via explicit prop passing) it is totally fine to inline.

---

The bug-or-feature here is that because React components are just functions the same considerations apply as for _any_ function that is defined inside another function. It is going to close over the variables in its scope, which means you can access them without having to pass them explicitly. On the plus side this might make it easier to use the components. On the negative side, you might wind up closing over "stale" state which could lead to a bug. Whether the closure helps or hurts you depends on the context.

And pretty much exact same deal for performance. By putting the function inside the other component it is getting recreated on every render. It will be short-lived and get garbage-collected, which could be wasteful. By hoisting it outside to the module root you have exactly one copy of it that gets reused for the lifetime of the program. Does this matter in practice? The answer is "probably not", but if you can hoist the function up without sacrificing too much (ie. because you don't really need to pass things via closure) then you should.

Reply:
Thanks for the advice @wincent